content/en/blog/_posts/2023-12-15-volume-attributes-class/index.md

```diff
@@ -0,0 +1,178 @@
+---
+layout: blog
+title: "Kubernetes 1.29: VolumeAttributesClass for Volume Modification"
+date: 2023-12-15
+slug: kubernetes-1-29-volume-attributes-class
+---
+
+**Author**: Sunny Song (Google)
+
+The v1.29 release of Kubernetes introduced an alpha feature to support modifying a volume
+by changing the `volumeAttributesClassName` that was specified for a PersistentVolumeClaim (PVC).
+With the feature enabled, Kubernetes can handle updates of volume attributes other than capacity.
+Allowing volume attributes to be changed without managing it through different
+provider's APIs directly simplifies the current flow.
+
+You can read about VolumeAttributesClass usage details in the Kubernetes documentation 
+or you can read on to learn about why the Kubernetes project is supporting this feature.
+
+
+## VolumeAttributesClass
+
+The new `storage.k8s.io/v1alpha1` API group provides two new types:
+
+**VolumeAttributesClass**
+
+Represents a specification of mutable volume attributes defined by the CSI driver.
+The class can be specified during dynamic provisioning of PersistentVolumeClaims,
+and changed in the PersistentVolumeClaim spec after provisioning.
+
+**ModifyVolumeStatus**
+
+Represents the status object of `ControllerModifyVolume` operation.
+
+With this alpha feature enabled, the spec of PersistentVolumeClaim defines VolumeAttributesClassName
+that is used in the PVC. At volume provisioning, the `CreateVolume` operation will apply the parameters in the
+VolumeAttributesClass along with the parameters in the StorageClass.
+
+When there is a change of volumeAttributesClassName in the PVC spec,
+the external-resizer sidecar will get an informer event. Based on the current state of the configuration,
+the resizer will trigger a CSI ControllerModifyVolume.
+More details can be found in [KEP-3751](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/README.md).
+
+## How to use it
+
+If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate
+in the `kube-controller-manager` and the `kube-apiserver`. Use the `--feature-gates` command line argument:
+
+
+```
+--feature-gates="...,VolumeAttributesClass=true"
+```
+
+
+It also requires that the CSI driver has implemented the ModifyVolume API.
+
+
+### User flow
+
+If you would like to see the feature in action and verify it works fine in your cluster, here's what you can try:
+
+
+1. Define a StorageClass and VolumeAttributesClass
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+   name: csi-sc-example
+provisioner: pd.csi.storage.gke.io
+parameters:
+   disk-type: "hyperdisk-balanced"
+volumeBindingMode: WaitForFirstConsumer
+```
+
+
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: VolumeAttributesClass
+metadata:
+   name: silver
+driverName: pd.csi.storage.gke.io
+parameters:
+   provisioned-iops: "3000"
+   provisioned-throughput: "50"
+```
```

Comment:
Please indent these lines by 3 spaces.

Reply:
I am not talking about the indentation of the YAML. I am talking about the Markdown that contains YAML snippets.

Reply:
This is too difficult to communicate via text, I will just type the example before I make any changes, Do you mean we need to change to:

```yaml
   apiVersion: storage.k8s.io/v1alpha1
   kind: VolumeAttributesClass
   metadata:
      name: silver
   driverName: pd.csi.storage.gke.io
   parameters:
      provisioned-iops: "3000"
      provisioned-throughput: "50"
```

with each line indent with 3 spaces as the starting point?

Reply:
(I'm explaining how to make an ordered list in Markdown - you indent everything until the start of the next list item)

Reply:
If this still isn't clear, I can push a commit with the required indentation. Let me know if I should.